city.lua

```diff
@@ -1325,6 +1326,24 @@ local function clearCell(city, upgradeCell)
     end
 end
 
+--- @param city City
+--- @param cellCoordinates Coordinates
+local function freeCell(city, cellCoordinates)
+    local cell = GridUtil.safeGridAccess(city, cellCoordinates)
+    if cell == nil then
+        return
+    end
+    -- check for incorrect cell
+    assert(cell.type == "building")
+
+    local p = cellCoordinates
+    city.grid[p.y][p.x] = {type = "unused"}
+    -- this doesn't include actual position, so shift to the top-left
+    addBuildingLocations(city, {x = p.x - 1, y = p.y - 1})
+    -- TODO: can't we just add it like this?
+    --city.buildingLocationQueue.pushleft(p)
```

Comment:
`addBuildingLocations()` works nicely here.
i'd prefer just to push this location into the building queue, but couldn't understand these code paths fully.

maybe i need to try some call visualizer -- still don't get it after reading files so many times :(

Reply:
Did you try this?

```
Queue.pushleft(city.buildingLocationQueue, coordinates)
```

Unfortunately the methods from `queue.lua` are only available via the `Queue` call, not on the field that we see as a queue.

See this as an example:

https://github.com/bahrmichael/factorio-tycoon/blob/49c2789f9e55dd1b441b868c45d08d4e4d3f4880/city.lua#L699

Reply:
i've tried this, but it **just forces** to build in that position. so, when it is called somewhere inbetween house upgrade - it just overlaps with second entity in that cell :(
also, i can't find proper place to add a check if that position in `city.buildingLocationQueue` is already occupied (for any reason)

ofc, i moved `script.register_on_entity_destroyed(entity)` to register most of our city entities (like gardens), otherwise nothing will be able to clear cells at runtime

Reply:
> i've tried this, but it just forces to build in that position. so, when it is called somewhere inbetween house upgrade - it just overlaps with second entity in that cell :(
also, i can't find proper place to add a check if that position in city.buildingLocationQueue is already occupied (for any reason)

I think I just forced so far, and didn't build safety checks down the line.

When you add an entry to a location queue, the code should eventually run through city#startConstruction. A few lines in there is a piece of code that removes duplicates from the queue. Iirc this was a bandaid that worked well enough to make it to production :D

https://github.com/bahrmichael/factorio-tycoon/blob/4579b5de58fa6cb04374a20d2a50cc914a132813/city.lua#L829-L831

To check if a location is already occupied, I would check the city's grid, translate the grid to map position and then search for entities there (if the city's grid has not been updated with an on_destroy hook).

I hope that helps, let me know if you have more questions!

Reply:
oh, that's why! i was searching for `buildingLocationQueue`, but here it is
```
coordinates = Queue.popleft(city[queueIndex])
```
with `queueIndex = "buildingLocationQueue"`, missed that. thanks!